// keep-a-changelog/script/deps/deno.land/std@0.205.0/semver/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NONE = exports.ALL = exports.ANY = exports.INVALID = exports.MIN = exports.MAX = void 0;
/**
 * MAX is a sentinel value used by some range calculations.
 * It is equivalent to `∞.∞.∞`.
 */
exports.MAX = {
    major: Number.POSITIVE_INFINITY,
    minor: Number.POSITIVE_INFINITY,
    patch: Number.POSITIVE_INFINITY,
    prerelease: [],
    build: [],
};
/**
 * The minimum valid SemVer object. Equivalent to `0.0.0`.
 */
exports.MIN = {
    major: 0,
    minor: 0,
    patch: 0,
    prerelease: [],
    build: [],
};
/**
 * A sentinel value used to denote an invalid SemVer object
 * which may be the result of impossible ranges or comparator operations.
 * @example
 * ```ts
 * import { eq } from "https://deno.land/std@$STD_VERSION/semver/eq.ts";
 * import { parse } from "https://deno.land/std@$STD_VERSION/semver/parse.ts";
 * import { INVALID } from "https://deno.land/std@$STD_VERSION/semver/constants.ts"
 * eq(parse("1.2.3"), INVALID);
 * ```
 */
exports.INVALID = {
    major: Number.NEGATIVE_INFINITY,
    minor: Number.POSITIVE_INFINITY,
    patch: Number.POSITIVE_INFINITY,
    prerelease: [],
    build: [],
};
/**
 * ANY is a sentinel value used by some range calculations. It is not a valid
 * SemVer object and should not be used directly.
 * @example
 * ```ts
 * import { eq } from "https://deno.land/std@$STD_VERSION/semver/eq.ts";
 * import { parse } from "https://deno.land/std@$STD_VERSION/semver/parse.ts";
 * import { ANY } from "https://deno.land/std@$STD_VERSION/semver/constants.ts"
 * eq(parse("1.2.3"), ANY); // false
 * ```
 */
exports.ANY = {
    major: Number.NaN,
    minor: Number.NaN,
    patch: Number.NaN,
    prerelease: [],
    build: [],
};
/**
 * A comparator which will span all valid semantic versions
 */
exports.ALL = {
    operator: "",
    semver: exports.ANY,
    min: exports.MIN,
    max: exports.MAX,
};
/**
 * A comparator which will not span any semantic versions
 */
exports.NONE = {
    operator: "<",
    semver: exports.MIN,
    min: exports.MAX,
    max: exports.MIN,
};
