// keep-a-changelog/script/bin.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("./_dnt.polyfills.js");
const dntShim = __importStar(require("./_dnt.shims.js"));
const mod_js_1 = require("./deps/deno.land/std@0.189.0/path/mod.js");
const mod_js_2 = require("./mod.js");
const mod_js_3 = require("./deps/deno.land/std@0.189.0/flags/mod.js");
const mod_js_4 = require("./deps/deno.land/x/ini@v2.1.0/mod.js");
const settings_js_1 = __importDefault(require("./src/settings.js"));
const argv = (0, mod_js_3.parse)(dntShim.Deno.args, {
    default: {
        file: "CHANGELOG.md",
        format: "compact",
        release: null,
        create: null,
        url: null,
        https: true,
        quiet: false,
        head: null,
    },
    string: ["file", "format", "url", "head"],
    boolean: ["https", "init", "latest-release", "quiet"],
});
const file = (0, mod_js_1.join)(dntShim.Deno.cwd(), argv.file);
try {
    if (argv.init) {
        const changelog = new mod_js_2.Changelog("Changelog").addRelease(new mod_js_2.Release("0.1.0", new Date(), "First version"));
        changelog.format = argv.format;
        save(file, changelog, true);
        dntShim.Deno.exit(0);
    }
    const changelog = (0, mod_js_2.parser)(dntShim.Deno.readTextFileSync(file));
    changelog.format = argv.format;
    if (argv["latest-release"]) {
        const release = changelog.releases.find((release) => release.date && release.version);
        if (release) {
            console.log(release.version?.toString());
        }
        dntShim.Deno.exit(0);
    }
    if (argv.release) {
        const release = changelog.releases.find((release) => {
            if (release.date) {
                return false;
            }
            if (typeof argv.release === "string") {
                return !release.version || argv.release === release.version.toString();
            }
            return !!release.version;
        });
        if (release) {
            release.date = new Date();
            if (typeof argv.release === "string") {
                release.setVersion(argv.release);
            }
        }
        else {
            console.error("Not found any valid unreleased version");
            dntShim.Deno.exit(1);
        }
    }
    if (argv.create) {
        const version = typeof argv.create === "string" ? argv.create : undefined;
        changelog.addRelease(new mod_js_2.Release(version));
    }
    save(file, changelog);
}
catch (err) {
    console.error(red(err.message));
    if (!argv.quiet) {
        dntShim.Deno.exit(1);
    }
}
function save(file, changelog, isNew = false) {
    changelog.url = argv.url || changelog.url || getRemoteUrl(argv.https);
    if (!changelog.url) {
        console.error(red('Please, set the repository url with --url="https://github.com/username/repository"'));
        changelog.url = "https://example.com";
    }
    if (changelog.url) {
        const settings = (0, settings_js_1.default)(changelog.url);
        if (settings) {
            changelog.head = settings.head;
            changelog.tagLinkBuilder = settings.tagLink;
        }
    }
    if (argv.head) {
        changelog.head = argv.head;
    }
    dntShim.Deno.writeTextFileSync(file, changelog.toString());
    if (isNew) {
        console.log(green("Generated new file"), file);
    }
    else {
        console.log(green("Updated file"), file);
    }
}
function red(message) {
    return "\u001b[" + 31 + "m" + message + "\u001b[" + 39 + "m";
}
function green(message) {
    return "\u001b[" + 32 + "m" + message + "\u001b[" + 39 + "m";
}
function normalizeUrl(url, https) {
    // remove .git suffix
    url = url.replace(/\.git$/, "");
    // normalize git@host urls
    if (url.startsWith("git@")) {
        url = url.replace(/^git@([^:]+):(.*)$/, (https ? "https" : "http") + "://$1/$2");
    }
    // remove trailing slashes
    url = url.replace(/\/+$/, "");
    return new URL(url);
}
function getRemoteUrl(https = true) {
    try {
        const file = (0, mod_js_1.join)(dntShim.Deno.cwd(), ".git", "config");
        const content = dntShim.Deno.readTextFileSync(file);
        const data = (0, mod_js_4.parse)(content);
        const url = data?.['remote "origin"']?.url;
        if (!url) {
            return;
        }
        return normalizeUrl(url, https).href;
    }
    catch (err) {
        console.error(red(err.message));
        // Ignore
    }
}
