// keep-a-changelog/script/src/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.settings = void 0;
/**
 * Temporary URLPattern shim
 * @see https://github.com/denoland/dnt/issues/336
 */
class URLPatternShim {
    constructor(pattern) {
        Object.defineProperty(this, "pattern", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        if (typeof pattern === "string") {
            pattern = pattern.replace(/\*/g, ".*");
            pattern = new RegExp(`^${pattern}$`);
        }
        this.pattern = pattern;
    }
    test(url) {
        return this.pattern.test(url);
    }
}
function getSettingsForURL(url) {
    return exports.settings.find((settings) => settings.pattern.test(url));
}
exports.default = getSettingsForURL;
exports.settings = [
    {
        pattern: new URLPatternShim("https://github.com/*"),
        head: "HEAD",
        tagLink(url, tag, previous) {
            if (!previous) {
                return `${url}/releases/tag/${tag}`;
            }
            return `${url}/compare/${previous}...${tag}`;
        },
    },
    {
        pattern: new URLPatternShim("https://gitlab.*/*"),
        head: "master",
        tagLink(url, tag, previous) {
            if (!previous) {
                return `${url}/-/tags/${tag}`;
            }
            return `${url}/-/compare/${previous}...${tag}`;
        },
    },
];
